### PR TITLE
Source code injection and linting

### PR DESCRIPTION
**Pull request by ngscheurich**
I think we'd benefit from having a linting tool set up so that all of our sample code retains a consistent style. Following that train of thought a bit, I also thought it would be nice to be able to edit example code with correct syntax highlighting and all of the other niceties a well-configured text editor affords. So, we come now to…

## Source code injection

Editing each bit of sample code in its own file seemed like something that would fit the bill and the Pandoc-powered build tool I put together seemed like a good place to hook something up to handle this. After wiring everything up, I ended up with a `code` directory to hold all the sample code and was able to change stuff like this:

> ````ruby
> require "curses" # require the curses gem
> include Curses   # mixin curses
>
> # The next three methods are provided by including the Curses module.
>
> init_screen      # starts curses visual mode
> getch            # reads a single character from stdin
> close_screen     # closes the ncurses screen
> ```

To stuff like this:

> `!{code/2_1_title_screen_1.rb}`

I didn't swap out the code in all the files but did enough to hopefully provide a solid example of the general workflow.

Cool. Now sample code is super easy to lint, which brings us to…

## Linting

RuboCop is the de facto linter for maintaining Ruby code style, so I figured that would be the way to go for this project. In order to avoid bikeshedding (including of the personal variety), I simply copied down [thoughtbot's RuboCop config](https://github.com/thoughtbot/guides/blob/master/style/ruby/.rubocop.yml), updated it for the most current version of RuboCop, and fixed a small typo. You can now simply run `rubocop` in the root directory to have the files in `example_app` and `code` linted (I've not yet fixed any violations).

### CODE REVIEW

**Comment by Trevoke** [comment]
That is a massive change. I am impressed by the amount of work you put into it. 

I've found that beyond the three great rules of Ruby ("indent is two spaces", "parens around every function call and method definition" and "snake_case all names except module names"), there's not much linting necessary in Ruby, usually.

When writing a book like this, it's usually better to have the code inlined, so that we have the continuity of the code and the text to organize our thoughts, so this particular change may make our end goal more difficult to achieve.


We should make sure that we are writing a book and not writing code instead of writing a book :)


I think you might be doing a lot of engineering where simpler solutions might be better fits. For instance, we could just convert the entire book to org-mode and/or pandoc, both formats that have plugins which would support embedded code blocks and, likely, things like indentation and such. (alternatively, it's possible that this book is already written in pandoc following pandoc's markdown: http://pandoc.org/MANUAL.html#pandocs-markdown )


I saw that you are a vim user, so this plugin might do what you want: https://github.com/vim-pandoc/vim-pandoc-syntax

**Comment by sardaukar** [comment]
I'm also of the opinion that inline code is quite important, and that linting is not a top priority. I've read books with peculiar code styles that were nonetheless easy to follow due to the accompanying text.

Now that we have epub output, let's try to focus more on the rest of the book in my opinion :)

**Comment by ngscheurich** [comment]
@Trevoke Thanks very much for your thoughtful input. Firstly, I feel like I do tend to be guilty of premature optimization more often than I'd care to admit, so I'm glad you're keeping me in check!

As for the linting, I was less concerned about things like indentation and casing and more about things like whether we write an array of symbols `[:like, :this]` or `%i[like this]`. *However*, that likely won't end up mattering much and—as you and @sardaukar so rightly pointed out—we should focus on writing a book rather than a codebase.

I agree with you that the cognitive overhead of switching between prose and code is a non-negligible concept that I didn't much account for, and is the biggest thing that makes this PR a non-starter.

I introduced a Pandoc-based compilation process in #33, so source code blocks are no problem as-is, and I'm cool with keeping things that way. Upon further reflection, I think one of the biggest pain points I was trying to address with these changes was proper syntax highlighting/formatting within said code blocks, and it looks like the Vim plugin you pointed me towards will take care of that.

**tl;dr** I agree with the @Trevoke and @sardaukar that these changes are not necessary, or at least premature, so I am closing this PR.
